capa/rules.py

```diff
@@ -1035,3 +1136,52 @@ def filter_rules_by_meta(self, tag: str) -> "RuleSet":
                     rules_filtered.update(set(capa.rules.get_rules_and_dependencies(rules, rule.name)))
                     break
         return RuleSet(list(rules_filtered))
+
+    def match(self, scope: Scope, features: FeatureSet, va: int) -> Tuple[FeatureSet, ceng.MatchResults]:
+        """
+        match rules from this ruleset at the given scope against the given features.
+
+        this routine should act just like `capa.engine.match`,
+        except that it may be more performant.
+        """
+        if scope == scope.FILE:
+            easy_rules_by_feature = self._easy_file_rules_by_feature
+            hard_rule_names = self._hard_file_rules
+        elif scope == scope.FUNCTION:
+            easy_rules_by_feature = self._easy_function_rules_by_feature
+            hard_rule_names = self._hard_function_rules
+        elif scope == scope.BASIC_BLOCK:
+            easy_rules_by_feature = self._easy_basic_block_rules_by_feature
+            hard_rule_names = self._hard_basic_block_rules
+        else:
+            raise Exception("programming error: unexpected scope")
+
+        candidate_rule_names = set()
+        for feature in features:
+            easy_rules = easy_rules_by_feature.get(feature)
+            if easy_rules:
+                candidate_rule_names.update(easy_rules)
+
+        # first, match against the set of rules that have at least one
+        # feature shared with our feature set.
+        candidate_rules = [self.rules[name] for name in candidate_rule_names]
+        features2, easy_matches = ceng.match(candidate_rules, features, va)
+
+        # now, match against (topologically ordered) list of rules
+        # that we can't really make any guesses about.
+        # these are rules with hard features, like substring/regex/bytes and match statements.
+        hard_rules = [self.rules[name] for name in hard_rule_names]
+        features3, hard_matches = ceng.match(hard_rules, features2, va)
```

Comment:
does this assume the features always overlap? since this is working on `features2`

having some trouble following along here

Reply:
`match` returns `(updated features, matches)`. the "updated features" are all the original features, plus added features from rule/namespace matches. so, we want to use `features2` instead of `features` because it will contain matches for rules like `read file`, etc. and a hard rule might be looking for those matches.

likewise, `features3` is a superset of `features2` is a superset of `features`. `features3` contains all original features plus all the match features, from both easy and hard rules. this is what we return to the caller.

Reply:
i'll update the inline documentation here to explain the above

Reply:
we could also assign `features` back to itself, but using the copies is a little more explicit about how we're tracking the data.

Reply:
2bf05ac6310c4a78fdac1cb26db7cc14134540a1